### PR TITLE
[WIP] work on #541

### PR DESCRIPTION
**Pull request by Vertmo**
Before your pull request is reviewed and merged, please ensure that:

* [X] there are no linting errors -- `npm run lint`
* [X] your code is in a uniquely-named feature branch and has been rebased on top of the latest master. If you're asked to make more changes, make sure you rebase onto master then too!
* [X] your pull request is descriptively named and links to an issue number, i.e. `Fixes #123`
* [ ] all the changes you wanted to do are made

I started work on #541 (adding API keys to a user profile). I already worked on the design of the "Advanced Settings" page, which you can find by clicking on "Advanced Settings" in the `/account` page (or by navigating to `/account/advanced`).
As you can see by the mockup, users won't be able to see keys on their interface (since we'll store them hashed), so the only time they'll be able to copy them will be at creation (I took inspiration in GitHub's Personal Access Token system).
Next, I'll be working on adding this keys to the database, and to add the logic to generate, hash them, and store them.

### CODE REVIEW

**Comment by Vertmo** [comment]
I'm still working on the key generation and display on the front-end, so take your time ^^

**Comment by Vertmo** [comment]
Now, this is starting to look good (in my opinion). Take your time !

File: client/modules/User/actions.js [hunk 1]
```diff
@@ -214,3 +215,33 @@ export function updateSettings(formValues) {
...
 }
+
+export function addApiKey(label) {
+  return ((dispatch) => {
+    crypto.randomBytes(20, (err, buf) => {
+      const key = buf.toString('hex');
+      const hashedKey = Buffer.from(key).toString('base64');
+      axios.put(`${ROOT_URL}/account/api-keys`, { label, hashedKey }, { withCredentials: true })
+        .then((response) => {
+          window.alert(`Here is your key :\n${key}\nNote it somewhere, you won't be able to see it later !`);
```
[the remark this answers is not in the document]
Comment (Vertmo) [reply client/modules/User/actions.js]: I'd like to use something similar to the github design too,m but I'm having doubts on how to do that exactly with React. The auto-file download seems like a good idea too. I'll look into that.

File: client/modules/User/pages/AdvancedSettingsView.jsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (Vertmo) [reply client/modules/User/pages/AdvancedSettingsView.jsx]: The thing is, I don't think we should overwhelm "normal" users who are only interested in writing a few sketches with some advanced options like this; that's why I wanted to hide such advanced options on another page. But your guess is as good as mine.

**Comment by Vertmo** [comment]
Basic auth seems like the best way to do it yes, but we should also think of how the API will end up evolving. Right now we're only planning on using it for sketch uploads right ? Do you have an idea of which other features will be available through the API later on ?

**Comment by catarak** [comment]
For API features, I'm thinking pretty basic CRUD for sketches. The goal is for the web editor to be able to interface with other platforms, e.g. a user could have their code on the web editor and GitHub, could upload examples they have locally on their computer easily.

**Comment by Vertmo** [comment]
That makes sense. Should I'll get working on the basic authentification as soon as I can !

**Comment by Vertmo** [comment]
Okay, I have added a Strategy for Basic Auth using passport-http. Changes dans fixes asides, I think this is starting to look correct. Tell me what you think ^^

**Comment by catarak** [comment]
i'm not sure how to test if the basic authentication with the key is working. do you have a test route?

**Comment by Vertmo** [comment]
Locally yes, I added : 
`app.get('/temp', passport.authenticate('basic', { session: false }), (req, res) => res.json(req.user));` to `server/server.js`

**Comment by Vertmo** [comment]
@catarak Is there anything I can do to further progress on that front ?

**Comment by catarak** [comment]
sorry for the delay, i'm not sure! i have a bit of a backup of PR's and i'm trying to get through some of those right now.

**Pull request closed without merging**